File: 2017/test/day_16_test.exs
defmodule AdventOfCode2017.Day16Test do
  use ExUnit.Case

  import AdventOfCode2017.Day16
  import AdventOfCode2017.TestHelper

  test "part1" do
    assert part1(sample_input_1(), "abcde") == "baedc"
    assert part1(input("test/input/day_16.txt")) == "pkgnhomelfdibjac"
  end

  test "spin" do
    assert spin(["a", "b", "c", "d", "e"], 3) == ["c", "d", "e", "a", "b"]
    assert spin(["a", "b", "c", "d", "e"], 1) == ["e", "a", "b", "c", "d"]
  end

  test "exchange" do
    assert exchange(["e", "a", "b", "c", "d"], 3, 4) == ["e", "a", "b", "d", "c"] 
  end

  test "partner" do
    assert partner(["e", "a", "b", "d", "c"], "e", "b") == ["b", "a", "e", "d", "c"] 
  end

  # test "part2" do
  #   assert part2(sample_input_1()) == 309
  #   assert part2(input("test/input/day_15.txt")) == 343
  # end

  test "parse" do
    assert parse(sample_input_1()) == [{"s", 1}, {"x", 3, 4}, {"p", "e", "b"}]
  end

  defp sample_input_1, do: "s1,x3/4,pe/b"
end

File: 2018/test/day_16_test.rb
require "test_helper"
require "day_16"

class Day16Test < Minitest::Test
  def setup
    @day_16 = Day16.new
  end

  def test_part1
    assert_equal 1, @day_16.part1(sample_input_1)
    assert_equal 590, @day_16.part1(input("day_16_1.txt"))
  end

  def test_part1
    assert_equal 475, @day_16.part2(input("day_16_1.txt"), input("day_16_2.txt"))
  end

  def test_parse
    assert_equal [[[3, 2, 1, 1], [9, 2, 1, 2], [3, 2, 2, 1]]], @day_16.parse(sample_input_1)
    assert_equal [[[3, 3, 2, 3], [3, 1, 2, 2], [3, 3, 2, 3]], [[1, 3, 0, 1], [12, 0, 2, 3], [1, 3, 0, 0]]], @day_16.parse(input("day_16_1.txt")).first(2)
  end

  def sample_reg_1
    [3, 2, 1, 1]
  end

  def sample_input_1
    "Before: [3, 2, 1, 1]
    9 2 1 2
    After:  [3, 2, 2, 1]"
  end
end
